### PR TITLE
feature: support logical view

### LINKED ISSUE #4610: Feature:  Support View Create/Select/Alter
**Summary**

Description for this feature.

`View` is kind of like UDS (user-defined-SQL), eg:

```
CREATE VIEW AS Stmt
```

`Stmt` is a query statement.

Example:

```
create table tmp_view as select number % 3 as a, avg(number) from numbers(1000) group by a;
```

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

* support logical view:

code base:
- [x] create
- [x] alter
- [x] drop
- [x] read from view

docs
- [x] docs

test
- [x] stateless-test

## Changelog

- New Feature

## Related Issues

Fixes #4610 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: query/src/interpreters/interpreter_view_create.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        .await?;
+
+        // check whether view has exists
+        if !self
+        .ctx
+        .get_catalog()
+        .list_tables(&*self.plan.tenant, &*self.plan.db)
+        .await?
+        .iter()
+        .all(|table| table.name() != self.plan.viewname.as_str()) {
```
Comment (flaneur2020) [inline query/src/interpreters/interpreter_view_create.rs:72]: try `any` instead? `any` might short-circuit the loop if find any matched condition without loop all the items 🤔

**Comment by sundy-li** [comment]
Almost looks good to me, missing some stateless tests .

**Comment by Veeupup** [comment]
> Almost looks good to me, missing some stateless tests .

ok, I'll write some stateless tests when alter is finished.

**Comment by b41sh** [comment]
need to add some docs for `View`.

File: tests/suites/0_stateless/05_ddl/05_0019_ddl_create_view.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+DROP VIEW IF EXISTS tmp_view;
+DROP VIEW IF EXISTS tmp_view2;
+CREATE VIEW tmp_view AS SELECT number % 3 AS a, avg(number) FROM numbers(1000) GROUP BY a ORDER BY a;
+CREATE VIEW tmp_view AS SELECT 1; -- {ErrorCode 2306}
+CREATE VIEW tmp_view2 AS SELECT * FROM numbers(100);
+DROP VIEW IF EXISTS tmp_view;
+DROP VIEW IF EXISTS tmp_view2;
```
Comment (sundy-li) [inline tests/suites/0_stateless/05_ddl/05_0019_ddl_create_view.sql:7]: New line needed

File: tests/suites/0_stateless/05_ddl/05_0021_ddl_alter_view.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+DROP VIEW IF EXISTS tmp_view;
+CREATE VIEW tmp_view AS SELECT number % 3 AS a, avg(number) FROM numbers(1000) GROUP BY a ORDER BY a;
+SELECT * FROM tmp_view;
+ALTER VIEW tmp_view AS SELECT number from numbers(3) ORDER BY number;
+SELECT * FROM tmp_view;
+DROP VIEW tmp_view;
```
Comment (sundy-li) [inline tests/suites/0_stateless/05_ddl/05_0021_ddl_alter_view.sql:6]: ditto

**Comment by leiysky** [comment]
What will happen when a table depended by a view get dropped?

**Comment by Veeupup** [comment]
> 

it acts as Clickhouse, `view` just converts query from view table into subquery.
so, if this happens, it'll give error messages just like Clickhouse to tell that the subquery's table not exists. cc @leiysky 
```shell
# in clickhouse
:) create view view_t1 as select a from t1;
:) drop table t1;
:) select * from view_t1;

SELECT *
FROM view_test

Query id: d3101f4a-2f80-4662-96aa-d3295172bc45


0 rows in set. Elapsed: 0.006 sec.

Received exception from server (version 21.12.1):
Code: 60. DB::Exception: Received from 0.0.0.0:9000. DB::Exception: Table default.t1 doesn't exist. (UNKNOWN_TABLE)
```
```shell
# in databend
databend :) create view view_t1 as select * from t1;
databend :) drop table t1;
databend :) select * from view_t1;

SELECT *
FROM view_t1

Query id: 87fb4594-bd23-4c34-856f-f53f8e6a878b
....
0 rows in set. Elapsed: 0.012 sec.

Received exception from server (version 2021.5.0):
Code: 1025. DB::Exception: Received from 0.0.0.0:9000. DB:Exception. Unknown table 't1'. ()
```

**Comment by leiysky** [comment]
> it acts as Clickhouse, `view` just converts query from view table into subquery.

Make sense.

**Comment by Veeupup** [comment]
review resolved. cc @sundy-li  @b41sh

File: query/src/interpreters/interpreter_table_drop.rs [hunk 4]
```diff
@@ -59,6 +61,15 @@ impl Interpreter for DropTableInterpreter {
             )
             .await?;
 
+        if let Some(table) = &tbl {
+            if table.get_table_info().engine() == VIEW_ENGINE {
+                return Err(ErrorCode::UnexpectedError(format!(
+                    "{}.{} is VIEW, planse use `DROP VIEW {}.{}`",
```
Comment (b41sh) [inline query/src/interpreters/interpreter_table_drop.rs:67]: typo `please`

File: query/src/sql/statements/statement_alter_view.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        ctx: Arc<QueryContext>,
+        table_name: &ObjectName,
+    ) -> Result<(String, String)> {
+        let idents = &table_name.0;
+        match idents.len() {
+            0 => Err(ErrorCode::SyntaxException("Create VIEW name is empty")),
+            1 => Ok((ctx.get_current_database(), idents[0].value.clone())),
+            2 => Ok((idents[0].value.clone(), idents[1].value.clone())),
+            _ => Err(ErrorCode::SyntaxException(
+                "Create VIEW name must be [`db`].`VIEW`",
```
Comment (b41sh) [inline query/src/sql/statements/statement_alter_view.rs:71]: Alter View

File: query/src/sql/statements/statement_alter_view.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    pub query: DfQueryStatement,
+}
+
+#[async_trait::async_trait]
+impl AnalyzableStatement for DfAlterView {
+    #[tracing::instrument(level = "debug", skip(self, ctx), fields(ctx.id = ctx.get_id().as_str()))]
+    async fn analyze(&self, ctx: Arc<QueryContext>) -> Result<AnalyzedResult> {
+        // check if query is ok
+        let _ = self.query.analyze(ctx.clone()).await?;
+        //
```
Comment (b41sh) [inline query/src/sql/statements/statement_alter_view.rs:45]: delete this useless comment line

File: query/src/sql/statements/statement_create_view.rs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    pub query: DfQueryStatement,
+}
+
+#[async_trait::async_trait]
+impl AnalyzableStatement for DfCreateView {
+    #[tracing::instrument(level = "debug", skip(self, ctx), fields(ctx.id = ctx.get_id().as_str()))]
+    async fn analyze(&self, ctx: Arc<QueryContext>) -> Result<AnalyzedResult> {
+        // check if query is ok
+        let _ = self.query.analyze(ctx.clone()).await?;
+        //
```
Comment (b41sh) [inline query/src/sql/statements/statement_create_view.rs:47]: ditto

File: query/src/sql/statements/statement_drop_view.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+}
+
+impl DfDropView {
+    fn resolve_table(&self, ctx: Arc<QueryContext>) -> Result<(String, String)> {
+        let DfDropView {
+            name: ObjectName(idents),
+            ..
+        } = self;
+        match idents.len() {
```
Comment (b41sh) [inline query/src/sql/statements/statement_drop_view.rs:59]: this `match` seems to appear many times, maybe we can rewrite it as a common function

File: website/versioned_docs/version-0.7.0/03-reference/03-sql/01-ddl/06-view/_category_.yml [hunk 9]
```diff
@@ -0,0 +1,4 @@
+label: 'Logical View'
```
Comment (b41sh) [inline website/versioned_docs/version-0.7.0/03-reference/03-sql/01-ddl/06-view/_category_.yml:1]: I think Version 0.7.0 is a released version and should not include the view.

Comment (BohuTANG) [reply website/versioned_docs/version-0.7.0/03-reference/03-sql/01-ddl/06-view/_category_.yml:1]: Yes, please don't put the new doc into the `version-0.7.0` folder, it's immutable after tagged.
I will tagging the new version this week with the command `yarn run docusaurus docs:version 0.7.[x]`, and your new documents will automatic in it soon.

Comment (Veeupup) [reply website/versioned_docs/version-0.7.0/03-reference/03-sql/01-ddl/06-view/_category_.yml:1]: get it

File: query/src/sql/statements/statement_drop_view.rs (hunk 8 above)
Comment (Veeupup) [reply query/src/sql/statements/statement_drop_view.rs:59]: make sense.
I make it by using the static method in `DfCreateTable::resolve_table`, it seems like a proper place to place this method

**Pull request merged**